mypy/checkexpr.py

```diff
@@ -2607,40 +2617,59 @@ def nonliteral_tuple_index_helper(self, left_type: TupleType, index: Expression)
             else:
                 return union
 
-    def _get_value(self, index: Expression) -> Optional[int]:
+    def _get_values(self, index: Expression) -> Optional[List[int]]:
         if isinstance(index, IntExpr):
-            return index.value
+            return [index.value]
         elif isinstance(index, UnaryExpr):
             if index.op == '-':
                 operand = index.expr
                 if isinstance(operand, IntExpr):
-                    return -1 * operand.value
+                    return [-1 * operand.value]
         typ = self.accept(index)
         if isinstance(typ, Instance) and typ.last_known_value is not None:
             typ = typ.last_known_value
         if isinstance(typ, LiteralType) and isinstance(typ.value, int):
-            return typ.value
+            return [typ.value]
+        if isinstance(typ, UnionType):
+            out = []
+            for item in typ.items:
+                if isinstance(item, LiteralType) and isinstance(item.value, int):
+                    out.append(item.value)
+                else:
+                    return None
+            return out
         return None
 
     def visit_typeddict_index_expr(self, td_type: TypedDictType, index: Expression) -> Type:
         if isinstance(index, (StrExpr, UnicodeExpr)):
-            item_name = index.value
+            key_names = [index.value]
         else:
             typ = self.accept(index)
-            if isinstance(typ, Instance) and typ.last_known_value is not None:
-                typ = typ.last_known_value
-
-            if isinstance(typ, LiteralType) and isinstance(typ.value, str):
-                item_name = typ.value
+            if isinstance(typ, UnionType):
+                key_types = typ.items
             else:
-                self.msg.typeddict_key_must_be_string_literal(td_type, index)
-                return AnyType(TypeOfAny.from_error)
+                key_types = [typ]
 
-        item_type = td_type.items.get(item_name)
-        if item_type is None:
-            self.msg.typeddict_key_not_found(td_type, item_name, index)
-            return AnyType(TypeOfAny.from_error)
-        return item_type
+            key_names = []
+            for key_type in key_types:
+                if isinstance(key_type, Instance) and key_type.last_known_value is not None:
+                    key_type = key_type.last_known_value
+
+                if isinstance(key_type, LiteralType) and isinstance(key_type.value, str):
+                    key_names.append(key_type.value)
+                else:
+                    self.msg.typeddict_key_must_be_string_literal(td_type, index)
+                    return AnyType(TypeOfAny.from_error)
```

Comment:
Would it be practical to refactor this, so that this uses `try_getting_str_literals()`? It looks like there is some code duplication. (If yes, the function should be probably moved to `nodes.py` and imported both here and in the plugin.)

Reply:
Unfortunately, no -- this segment of code permits unicode keys but the `try_getting_str_literals` section doesn't.

IMO the correct thing to do is (a) modify `try_getting_str_literals` so it also allows unicode keys and (b) convert this entire function into a plugin.

But I think making this change fully correct is partly blocked on https://github.com/python/mypy/issues/6123, so I decided to defer the refactor for now.